pahfit/fitter.py

```diff
@@ -0,0 +1,209 @@
+       passed during step 1. When fit result uncertainties are
+       implemented, they will also need to be retrieved through this
+       API.
+    4. Access to the evaluation of the underlying model (again with no
+       assumptions like in step 2.).
+
+    A few notes on how the above is achieved:
+
+    For the model setup, there is one function per type of component
+    supported by PAHFIT, and the arguments of these functions will ask
+    for certain standard PAHFIT quantities (in practice, these are the
+    values stored in the Features table). The abstract Fitter class
+    ensure that the function signatures are the same between different
+    Fitter implementations, so that only a single logic has to be
+    implemented to up the Fitter (in practice, this is a loop over the
+    Features table implemented in Model).
+
+    During the Fitter setup, the initial values, bounds, and "fixed"
+    flags are passed using one function call for each component, e.g.
+    register_line(). Once all components have been added, the
+    finalize_model() function should be called; some subclasses (e.g.
+    APFitter) need to consolidate the registered components to prepare
+    the model that they manage for fitting. After this, fit() can be
+    called to apply the model and the astropy fitter to the data. The
+    results will then be retrievable for one component at a time, by
+    passing the component name to get_result().
+
+    """
+
+    @abstractmethod
+    def clear(self):
+        """Reset model.
+
+        After reset, register_() and finalize_model() can be used again.
+
+        """
+        pass
+
+    @abstractmethod
+    def components(self):
+        """Return list of features.
+
+        Only works after finalize_model(). Will return the names passed
+        using the register functions.
+
+        """
+        pass
+
+    @abstractmethod
+    def finalize_model(self):
+        """Process the registered features and prepare for fitting.
+
+        The register functions below allow adding individual features.
+        The exact implementation of how features are added, and
+        finalized in to a single fittable model, depend on the
+        underlying implementation.
+
+        """
+        pass
+
+    @abstractmethod
+    def register_starlight(self, name, temperature, tau):
+        """Register a starlight feature.
+
+        The exact representation depends on the implementation, but the
+        meaning of the parameters should be equivalent.
+
+        Parameters
+        ----------
+        name : str
+            Unique name. Will be used to allow retrieval of the results
+            after the fitting.
+
+        temperature : array of size 3
+            Blackbody temperature, given as [value, lower_bound,
+            upper_bound]. The bounds assume the same system as the
+            features table: if they are masked, the parameter will be
+            fixed, while +inf or -inf mean unbounded.
```

Comment:
We have now updated the way we store bounds.  Worth transferring that (e.g. a scalar value (no bounds) or structured array with three fields)?  We should take `nan` and `mask` off the table: if we are spoon-feeding `Fitter` don't feed it anything masked, pre-translate `nan` to +-`inf`.

Reply:
Good idea. I was actually having trouble with this while trying to make the tests work. Right now there is still some logic that deals with the nans (to be fair they are still quite useful for determining if something is fixed based on the bounds given).

Reply:
That's for `Model` in our new way of thinking.

Reply:
We can indicate fixed by passing a scalar value in the spoon-feeding stage.

Reply:
Yes we should definitely deal with these details in `Model`. I will try to make it work with the scalar vs 3-array approach. That does feel less cryptic, so I made the change and pushed it. I implemented the array/scalar check using `__hasattr__(array_or_scalar, '__len__')`. Seems to run safely according to the tests.

Reply:
Inside your `Fitter` you mean?  Are you passing the val/min/max flavor when there are bounds?

Reply:
The `Fitter` setup logic (implemented in `Model._set_up_fitter()` cleans the values before they are passed to `add_feature_*()`, and will pass a scalar or a 3-array as appropriate. The above check happens inside `APFitter`, where it checks if it receive a scalar or not.